### DISCUSSION TITLE
Project primary keys for aggregate collection to array/list of integers

### DISCUSSION THREAD

**Discussion by dtc13579** [leonardoporro/Detached-Mapper#102] [Q&A]
I’m wanting to be able to project a query of an aggregated collection to an array or list of the integer primary keys.  Listed below are the options I’ve tried.  (I'd also like to use List<int> to apply the updated aggregate to the database.)

I get an error “Object reference not set to an instance of an object”.  I assume the library doesn’t natively support this?  (Or maybe I’m doing something wrong.)  

Can a custom extension be made to handle this?  If so, can you point me in the right direction?

Thanks

```
class Invoice
{
    public int Id { get; set; }

    [Aggregation]
    public List<Location> Locations { get; set; }

    [Composition]
    public List<InvoiceRow> Rows { get; set; }
}

//context.Project<Invoice, DTO1>(context.Invoices).ToList();  --> doesn't work
class DTO1
{
    public int Id { get; set; }
    public List<int> Locations { get; set; }
}

//context.Project<Invoice, DTO2>(context.Invoices).ToList();  --> doesn't work
class DTO2
{
    public int Id { get; set; }
    public int[] Locations { get; set; }
}

class Identity
{
    public int Id { get; set; }
}
//context.Project<Invoice, DTO_array_of_objects_works>(context.Invoices).ToList();  --> works
class DTO_array_of_objects_works
{
    public int Id { get; set; }
    public List<Identity> InvoiceTypes { get; set; }
}

```

**Comment by leonardoporro** [comment]
Hi Devon,
Right now, Project is a kind of draft feature, it does the basics but doesn't integrate or implement very well with the other mapper features.
For what you are trying to do, a custom mapping like locations.Select(l => l.Id).ToList() must be implemented which is not supported at this time.
I'll open an issue (or you can do it yourself if you want) to track this.
Your conribution to the library would be very very welcomed, but I'm not sure about how to point you in the right direction. I would start by reading the wiki and getting familiar about how all the mapping works.
Also, if you want to help me to speed up this feature, some tests would be great. Create a test with your expectation and submit a pull request to Detached.Mappers.EntityFramework.Tests so I can focus on how to make it work.
The test can also contains suggestions about how this feature would be configured. Of course, it doesn´t need to compile, and it probably won´t.
Thanks for your interest and feedback

**Comment by dtc13579** [comment]
Similarly, we'd also like to Map an array or list of integer Ids.
Is this also currently not supported?  (I get an “Object reference not set to an instance of an object” error)
Thanks

```
context.Map<Invoice>(new
            {
                Id = 1,

                InvoiceTypes = new[] { 1,2 },  
                //Also tried InvoiceTypes = new List<int>() { 3,2 },

                Rows = new[]
                {
                    new { Id = 1, Description = "prod 1", Quantity = 5, Price = 25 },
                    new { Id = 2, Description = "prod 2", Quantity = 3, Price = 100 }
                }
            });
```

**Comment by leonardoporro** [comment]
The mapper is entirely customizable, you can write your own mapping routines.
Here I added a [sample](https://github.com/leonardoporro/Detached-Mapper/blob/main/test/Detached.Mappers.EntityFramework.Tests/CustomMapperTests.cs).

I'm planning to extend it even more, by adding the posibility to specify a mapper, or lambda, to a specific member of a type pair.

On the other hand... I see that you are working with Keys and FKs.. Not sure if suits your needs but for these cases I usually create a type containing only the key, and call it Reference, Identity or something like that.

So that a DTO property can be a `Reference { int Id; } `  or a collection of References, and can be mapped to entities (let's say User { int Id; }) without adding any code or modification.

Please take a look at [Foreing Keys](https://github.com/leonardoporro/Detached-Mapper/wiki/Foreign-Keys) article on the wiki, maybe it helps. Also it's quite easy to adapt your frontend controls/forms to wrap Ids in an object to make it work with this Reference pattern.

**Comment by leonardoporro** [comment]
Another clarification that might be oportune for this discussion. 

I made this library to quickly map DTOs (and even dictionaries) to Entities so people stops being afraid of creating the right objects for the corresponding layers.

The idea is to keep it in a convention-over-configuration way, matching by property name and mapping similiar kind of structures, while keeping it open for modification (so that you can inject your own mapper at any point). 

I'm trying to walk away from the path of AutoMapper, don't want to add so many mapping options and configurations because:
- It´s only me here and I don´t have resources to maintain tons of features
- It ends making the software slower
- People use the features in a wrong way, code starts failing and become difficult to debug, Architects starts hating the package and 
finally they remove it from all solutions. Like Automapper.

**Comment by leonardoporro** [comment]
There was a large refactor for 7.2 that allows me to implement this feature easily for both maps and projections

**Comment by leonardoporro** [comment]
Map keys to entities is out in 8.0.1.
[https://github.com/leonardoporro/Detached-Mapper/blob/main/test/Detached.Mappers.EntityFramework.Tests/KeyToEntityTests.cs](https://github.com/leonardoporro/Detached-Mapper/blob/main/test/Detached.Mappers.EntityFramework.Tests/KeyToEntityTests.cs)

I still need to make it work for projections and write the documentation.
Property must follow this convention: [original property name] + [foreign entity key name]
e.g.:  
user: { group: { id: } -> userDTO: { groupId: }

**Comment by leonardoporro** [comment]
Seems to work in both ways.. [here](https://github.com/leonardoporro/Detached-Mapper/blob/8dffa6dfdf3f4a475eb7f15de263f6e5a5c4c046/test/Detached.Mappers.EntityFramework.Tests/Features/BindForeingKeyToDto.cs) and [here](https://github.com/leonardoporro/Detached-Mapper/blob/main/test/Detached.Mappers.EntityFramework.Tests/Features/MapForeignKeyToEntity.cs)
Still fresh, need a lot of testing...
Thanks for the feedback
